feat: 알림 기능을 적용한다

devholic22 opened the pull request:
## 📄 Summary

* FCM 방식을 이용한 알림 시스템을 적용하였습니다. [참고 글](https://headf1rst.github.io/TIL/push-notification)
* 더 빠른 동작을 하기 위해 비동기 방식을 이용하였습니다.
* 다른 도메인에서 알림을 보내고자 할 경우에는 `AlertCreatedEvent`를 발생시키면 됩니다.
    ```java
    public record AlertCreatedEvent(
        AlertGroup group, // 알림 그룹 (좋아요, 알림 등)
        String title, // 알림 제목
        String body, // 알림 상세 내용 (내부 메시지 등)
        String sender, // 발신자의 닉네임
        Long receiverId // 수신자의 ID
    ) {
    }
    ```
* 회원에 대한 FCM 토큰이 필요합니다. 이것을 위해 회원이 로그인할 시 토큰 생성 이벤트 (AlertTokenCreatedEvent)를 발생시키도록 했습니다. (로그인 테스트에 이벤트 발생 검증을 적용하였습니다.)
* 회원이 로그아웃할 시 토큰을 없애야 하는데, 이 경우는 로그아웃 기능이 완료된 뒤 호출하면 될 것 같습니다.
* 회원의 토큰을 관리하는 것으로는 레디스를 이용하였습니다. (갱신/삭제 빈번, 토큰 데이터가 key-value 형식인 점 고려)
* 회원이 받은 알림을 단일 조회할 때는 읽음 여부가 업데이트 됩니다. AlertQueryService와 AlertService 중 어느 곳에 해당 메서드를 둘 지 고민했었는데, AlertQueryService는 순수한 조회, AlertService는 내부 작업이 포함된 모든 기능들을 관리하도록 하는 게 더 나을 것 같아 AlertService에 단일 조회를 두었습니다.
* 모든 알림은 생성된 지 60일이 초과되면 (즉, 61일 이상) 삭제 처리됩니다. 다만 통계를 위해 사용해야 할 수도 있으므로 soft delete를 이용하였습니다. 이때, soft delete를 하면서도 부모 (BaseEntity)의 값을 정해줘야 하기에 `@SuperBuilder`를 적용해주었습니다.
## 🙋🏻 More
### 1차 내용
* 파이어베이스 빈을 등록할 때 민감한 정보를 드러내지 않기 위해 `@Value`로 암호화된 값들을 가져오고, 이들을 Map으로 만든 뒤 바이트스트림 화 하는 것으로 문제를 해결했습니다. 특히 테스트 환경일 때 문제가 발생했는데, 다행히 파이어베이스에서 기본 GoogleCredentials을 발행해주는 코드가 있어 문제 발생 상황 (IOException, 테스트 환경일 때 발생)을 해결했습니다.
* 알림 보냄 여부를 테스트할 때 Fake 객체로 만들긴 했으나 (AlertFakeManager) 이 방법이 충분할 지 의문이 듭니다. 혹시 더 좋은 방법이 있을지 더 고민해보도록 하겠습니다.
* 로컬에서 테스트했을 때는 문제 없이 수행되다가 CI 환경에서만 실패했었는데, 이전에 호감 조회 기능을 개발했을 때에도 겪었던 이슈라 금방 해결할 줄 알았으나 오래 걸렸습니다. 그러다가 [LocalDateTime 정밀도 관련 문제](https://jun27.tistory.com/72)임을 알게 되었고, 다른 기능들을 개발할 때 (최신 순 정렬 조회가 필요한 경우 등) LocalDateTime을 쓰지 않고 시간 차이를 표현할 수 있는 다른 객체 (ex: 제목 1, 제목 2처럼 늘려가는 방법)만 검증하는 게 좋겠다는 생각이 들었습니다!
### 2차 내용 (분산 락)
[참고 문서 - 컬리 기술 블로그 '풀필먼트 입고 서비스팀에서 분산 락을 사용하는 방법'](https://helloworld.kurly.com/blog/distributed-redisson-lock/)
* 해결하려는 문제: 주기적으로 생성된 지 60일 이상 된 알림들을 삭제 상태로 변환시키는 스케줄러를 작성하였습니다. 그러나 서버가 확장되어 스케일 아웃 될 경우, 기존 방식대로라면 여러 인스턴스에서 같은 시간에 UPDATE SQL을 보내는 이슈가 발생할 수 있겠다는 것을 파악했습니다.
* 락 방식 선택 - synchronized / 낙관적 락 / 비관적 락 / `분산 락`
    * synchronized: 어플리케이션 단위에서 제어되는 기술이기 때문에, 본 문제와는 관련이 없다고 생각했습니다.
    * 낙관적 or 비관적 락: 낙관적 락이나 비관적 락은 데이터베이스의 특정 레코드에 대해서 사용되는 기법이지만, 본 문제는 60일 이상 된 알림들을 모두 UPDATE하는 쿼리를 다루고 있기 때문에 관련이 없다고 생각했습니다.
    * 이러한 이유로 문제에서 요구되는, 다수의 서버에서 발생하는 동시성을 해결하기 위해서는 분산 락을 이용해야겠다고 생각했습니다.
* 기술 선택 - MySQL / Zookeeper / `Redis`
    * MySQL을 이용한 분산 락 (네임드락)은 락을 사용하기 위해 별도의 커넥션 풀을 관리해야 하고, 락 부하를 RDB에서 받게 됩니다. 이미 Redis를 프로젝트에서 사용하기도 하고, 성능이 더 중요하다고 생각해 적용하지 않았습니다.
    * Zookeeper는 현재 프로젝트에서 Zookeeper를 사용하지 않으며, 학습 비용이 다른 두 가지에 비해 더 높다는 판단을 하여 적용하지 않았습니다.
    * Redis를 이용하면 인메모리 특성상 빠른 속도를 낼 수 있고, 별도의 설정이 필요하지만 Redis를 본 프로젝트에서 이미 사용한다는 점 때문에 Redis를 이용한 분산 락 방식이 스케줄러에 더 적합하다는 생각이 들었습니다.
* 방식 선택 - Lettuce / `Redisson`
    * Lettuce: Lettuce는 공식적으로 분산 락을 지원하지 않기에 직접 구현해야 합니다. 또한, 스핀 락 방식으로 진행되어 락을 획득하지 못할 경우 락 획득을 위해 지속적으로 요청해야 하는 점이 있어 최대한 부하를 줄이고자 선택하지 않게 되었습니다.
    * Redisson: Redisson을 이용한 분산 락은 RLock을 이용하며, 이것의 내부 코드를 보면 발행/구독 (pub/sub)으로 이루어져 있음을 볼 수 있습니다. 이 덕분에 스핀 락 방식에 비해 추가적으로 부하가 발생하지 않으며, 락 획득 시도/반환 등의 코드를 쉽게 제공해주어 Redisson을 선택하게 되었습니다.

close #34

devholic22 committed:
chore: Firebase 의존성 등록

devholic22 committed:
chore: Firebase 설정 파일 등록

- Firebase 설정 파일 환경 등록

devholic22 committed:
feat: 알림 도메인 기초 설계

- 알림 도메인 기초 설계 작성
- 관련 예외 핸들러 등록

devholic22 committed:
feat: 알림 전송 기능 구현

- 알림 전송 기능에 대한 서비스, 구현체 작성

devholic22 committed:
chore: Firebase key gitignore 등록

- 보안을 위해 Firebase key gitignore에 등록

devholic22 committed:
refactor: body NotEmpty 제약사항 제거

devholic22 committed:
feat: 알림에 AlertGroup 정보 추가

- 알림에 AlertGroup 정보 추가
- AlertManager 패키지 위치 수정

devholic22 committed:
feat: 알림에 발신자 정보 추가

- 알림에 발신자 정보 추가 (닉네임)
- 보낸 발신자가 없을 경우 시스템이 보낸 것으로 간주

devholic22 committed:
feat: 알림 이벤트 핸들러 추가

- 알림 이벤트 핸들러 추가
- CreateAlertRequest 삭제 대신 이벤트 인자로 변경

devholic22 committed:
feat: 알림 데이터베이스 저장 추가

- 보낸 알림과 데이터베이스 연결 작성

devholic22 committed:
refactor: Soft delete를 적용하기 위해 상속자 변경

- Soft delete를 적용하기 위해 SoftDeleteBaseEntity를 상속받도록 수정

devholic22 committed:
feat: 60일 초과된 알림은 soft delete되도록 처리

devholic22 committed:
feat: 알림 생성 시각 추가

- 알림 생성 시각 추가

devholic22 committed:
refactor: Alert 속성 변경

- 알림에 쓰인 토큰을 관리하지 않도록 제거
- 알림의 대상이 되는 회원의 ID가 연결되도록 수정

devholic22 committed:
Merge remote-tracking branch 'origin/develop' into feat/34

devholic22 committed:
feat: 레디스를 이용한 FCM 토큰 관리 기능 구현

- 로그인 시 토큰을 저장하도록 설정
- 토큰 관리 리포지터리 정의

devholic22 committed:
refactor: 알림 필드 수정

- senderId 대신 receiverId를 관리하도록 수정
- 토큰을 직접 받기보다 서비스 내부에서 가져오도록 수정

devholic22 committed:
refactor: 토큰 조회 시 유효성 검사 진행

- 토큰 조회 시 유효성 검사 진행

devholic22 committed:
feat: 알림 조회 기능 개발

- 알림 조회 기능 개발 (페이징, 단일)
- 알림 단일 조회 시 read 작업 수행되도록 설정

devholic22 committed:
test: 알림 도메인 테스트 작성

- 알림 도메인 단위테스트 작성

devholic22 committed:
test: AlertJdbcRepository 테스트 작성

- AlertJdbcRepository 테스트 작성
- 관련 Fixture 생성, 값을 작성하기 위해 빌더 패턴 추가

devholic22 committed:
test: AlertJpaRepository 테스트 작성

- AlertJpaRepository 테스트 작성
- 관련 Fixture 생성

devholic22 committed:
test: AlertQueryRepository 테스트 작성

- AlertQueryRepository 테스트 작성
- 페이징 조회 타입을 QueryResults로 변경
- AlertSearchResponse 반환 타입 변경
- 관련 Fixture 등록

devholic22 committed:
refactor: 파일 이름 변경

- 레디스 의존성을 드러내기 위해 파일 이름 변경

devholic22 committed:
test: FakeAlertTokenRepository 작성

- FakeAlertTokenRepository 작성

devholic22 committed:
test: FakeAlertRepository 작성

- FakeAlertRepository 작성

devholic22 committed:
test: AlertService 테스트 작성

- AlertService 테스트 작성
- 관련 Fixture 등록
- FakeAlertRepository 문제 수정

devholic22 committed:
fix: 알림 정렬 기준 문제 수정

- 알림 정렬 조건에 id 추가되도록 수정

devholic22 committed:
test: AlertQueryService 테스트 작성

- AlertQueryService 테스트 작성
- 관련 Fixture 등록

devholic22 committed:
fix: FirebaseApp 중복 문제 해결

- FirebaseApp 중복 문제 해결

devholic22 committed:
test: 단일 알림 조회 테스트 수정

- 단일 알림 조회 테스트 수정
- API 문서 작성

devholic22 committed:
test: AlertControllerAcceptance 테스트 작성

- AlertControllerAcceptance 테스트 작성
- 관련 Fixture 등록

devholic22 committed:
refactor: 회원 로그인 시 이벤트 발생 검증 추가

- 회원 로그인 시 FCM 토큰 저장 이벤트 발생 검증 추가
- 회원 로그인 문서에 FCM 토큰 정보 추가

devholic22 committed:
refactor: 알림 단일 조회 반환값 변경

- 알림 단일 조회 반환값 (읽음 여부) 변경

devholic22 committed:
refactor: 알림 조회 테스트 정렬 기준 교정

- 알림 조회 테스트 정렬 기준 교정

devholic22 committed:
refactor: 필요없는 인자 제거

- 필요없는 인자 제거

devholic22 committed:
refactor: 파이어베이스 키 조회 방식 수정

- 직접적인 파일을 미리 만들지 않고 암호화된 값을 통해 파일을 만드는 방식으로 조회하도록 수정

devholic22 committed:
fix: 파이어베이스 키 조회 방식 수정

- 테스트 환경에서의 문제 해결하기 위해 테스트 application.yml에 파이어베이스 관련 정보 등록
- 파일 생성 중단, 데이터만 이용하도록 변경
- 빈 생성 중 예외 발생 (테스트 환경)일 시, 기본 GoogleCredentials 반환하도록 수정

devholic22 committed:
refactor: Fake 객체 이름 변경

- Fake 객체 이름 규칙에 맞게 변경

devholic22 committed:
docs: 의미없는 gitignore 삭제

- 의미없는 gitignore 코드 삭제

devholic22 committed:
refactor: FirebaseFileNotFoundException 삭제

- FirebaseFileNotFoundException 파일 삭제

sosow0212 commented on build.gradle:
```diff
@@ -78,6 +78,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
+
+    // Firebase
+    implementation 'com.google.firebase:firebase-admin:8.0.0'
```
혹시 8.0.0으로 하신 이유가 있을까요? 가장 많이 사용하는 버전은 8.1이고 현재는 9버전까지 나와서 어떤 차이가 있는지 몰라서 선택하신 다른 이유가 있는지 궁금해서 여쭤봅니다!

devholic22 replied:
다른 자료들을 봤을 때 처음 접했던 버전이라 적용했었는데, 말씀하신 뒤 파이어베이스의 업데이트 내역을 보니 [sendEachAsync](https://github.com/firebase/firebase-admin-java/pull/815), [트래픽 처리 개선](https://github.com/firebase/firebase-admin-java/pull/900) 등 더 좋은 기능들이 나온 걸 확인했습니다. 9.3.0 버전으로 변경했습니다 :)

sosow0212 commented on src/main/java/com/atwoz/alert/config/FirebaseConfig.java:
```diff
@@ -0,0 +1,118 @@
+@Configuration
+public class FirebaseConfig {
+
+    private static final String TYPE = "type";
```
오우 상수화랑 밸류 담느라 고생하셨네요 ㅠ

sosow0212 commented on src/main/java/com/atwoz/alert/domain/Alert.java:
```diff
@@ -0,0 +1,77 @@
+@EqualsAndHashCode(of = "id", callSuper = false)
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Alert extends SoftDeleteBaseEntity {
```
이거 softDelete로 처리하신 이유가 있나요?

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/AlertQueryRepository.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final JPAQueryFactory jpaQueryFactory;
+
+    public Page<AlertSearchResponse> findMemberAlertsWithPaging(final Long memberId, final Pageable pageable) {
```
소개팅 어플에선 자주 사용하는 API가 될텐데 인덱싱도 적용해보면 좋을 것 같네요!

devholic22 replied:
Alert를 조회할 때 사용되는 where 쿼리가 모두 동등 쿼리 (`receiver_id = ? and deleted_at is null`)라는 점, order by에 사용되는 정렬 조건이 모두 DESC 내림차순이라는 점을 이용하여 `receiver_id, deleted_at, created_at DESC, id DESC`로 인덱스를 적용하였습니다. [참고한 글](https://jojoldu.tistory.com/481)

비교를 위해 더미 데이터 백만 건의 데이터를 넣어봤는데, 해당 인덱스를 이용하기 전에는 약 0.548초가 일어났으나 인덱스 적용 후에는 약 0.0025초로 개선할 수 있었습니다.

혹시 인덱스 컬럼 선택을 잘못한 것이라면 피드백 부탁드립니다!

sosow0212 commented on src/main/java/com/atwoz/alert/application/AlertService.java:
```diff
@@ -0,0 +1,52 @@
+        return targetAlert;
+    }
+
+    @Scheduled(cron = MIDNIGHT)
+    public void deleteExpiredAlerts() {
+        alertRepository.deleteExpiredAlerts();
+    }
```
스케줄링은 관리를 위해 다른 클래스를 사용하는 건 어떨까요?
- 스케줄 설정도 따로 하지 않아도 될까요? 스케일아웃 환경일 땐 문제가 없을지 궁금하네요!
- 받은 알림은 무조건 자정에 삭제가 되는지도 궁금하네요! 소개팅에서 채팅이 오는 경우도 자정만 되면 삭제될 것 같은데 다른 방법은 없을까요?

devholic22 replied:
* 말씀하신 대로 스케일 아웃 환경일 때에는 스케줄링 메서드를 수행함에 있어 중복 호출될 여지가 있겠다는 생각이 들어, 락 기법 중 레디스의 Redisson을 이용한 분산 락을 이용하였습니다.
* 소개팅에서 채팅이 오는 경우 자정이 되었을 때 삭제시키는 것은 아니고, `모든 알림들 중 생성된 지 60일이 지난 알림들의 상태를 deleted 상태로 업데이트하는 쿼리`라 말씀하신 경우는 해당되지 않는 것 같습니다..!
* 추가로 분산 락을 이용한 테스트 (`RedissonAlertSchedulerTest`)를 작성했는데, 고쳐야 할 게 있다면 다른 방식으로 작성하도록 해보겠습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/application/AlertEventHandler.java:
```diff
@@ -0,0 +1,24 @@
+
+    private final AlertService alertService;
+
+    @EventListener
```
@TransactionalEventListener와 해당 어노테이션은 무엇이 다를까요!?

devholic22 replied:
EventListener는 트랜잭션에 참여하지 않는 환경에서 쓰이고, TransactionalEventListener는 트랜잭션을 인식한 뒤 발생하도록 하는 데 쓰입니다. 
TransactionalEventListener를 이용하면 외부 트랜잭션이 커밋된 후 수행되도록 하는 등의 옵션을 줄 수 있어서, 아래 말씀해주신 상황 (서비스 트랜잭션 롤백)에서의 발생을 방지할 수 있겠다는 생각이 들어 TransactionalEventListener로 수정 완료하였습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/application/AlertService.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public Alert readAlert(final Long memberId, final Long id) {
+        Optional<Alert> alert = alertRepository.findByMemberIdAndId(memberId, id);
+        if (alert.isEmpty()) {
+            throw new AlertNotFoundException();
+        }
+        Alert targetAlert = alert.get();
```
orElseThrow로 축약이 가능해보입니다~

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/FirebaseAlertManager.java:
```diff
@@ -0,0 +1,40 @@
+                .putData(CREATED_TIME, time)
+                .build();
+        FirebaseMessaging firebase = FirebaseMessaging.getInstance();
+        firebase.sendAsync(firebaseMessage);
```
해당 코드가 PR에 적어주신 비동기 처리 부분 같은데 비동기 메서드는 어떻게 작동 될까요?
추가적으로 현재 이벤트가 비동기로 처리돼서 알람을 보냈는데, 서비스부 트랜잭션 커밋 전에 롤백이 발생되면 잘못된 알림이 갈 수도 있을 것 같아요.

제가 잘 모를 수 있어서 이 부분에 대해서 간단하게 코멘트 부탁드립니다!!

devholic22 replied:
* 이벤트 발행이 기존에 동기적이었다는 점, `@TransactionalEventListener`를 이용할 때 트랜잭션 전파 수준을 REQUIRES_NEW로 하면 심각할 경우 데드락이 발생할 수 있다는 점을 확인하여 이를 해결하기 위해 이벤트 단에 `@Async`를 적용하였습니다. [참고 글](https://seongonion.tistory.com/166?category=929702)
* 비동기 설정 파일 (AsyncConfig)에 비동기 전용 executor, 알림 콜백 전용 executor를 생성했습니다.
* 파이어베이스에서 알림을 호출 (send)할 때, 비동기-논블록킹 방식으로 진행되기 위해 addListener 메서드를 이용하였습니다. [참고 글](https://dd-developer.tistory.com/130) 더불어, 알림 콜백 전용 executor의 스레드 풀 크기는 10개로 설정하였습니다.
* 파이어베이스는 기본적으로 `Executors.newCachedThreadPool()` (요청이 n개면 n개의 스레드를 생성)하는 방식이기 때문에 OOM 이슈가 발생할 수 있다고 판단, 최대 40개의 스레드 제한을 가진 FirebaseThreadManager를 생성하였습니다.
* 알림을 재전송하는 로직을 추가하였습니다. 제한 시간은 파이어베이스에서 권장한 대로 1초, 2초, 4초 세 번을 시도하도록 하였습니다.

정리하자면 알림 비동기 메서드는 다음 과정으로 진행됩니다.
1. 외부 서비스에서 알림 전송 이벤트 (AlertCreatedEvent) 호출
2. 이벤트에 `@Async`가 붙어 있기 때문에 기존 스레드 흐름과 별도로 분리된 스레드 생성됨 - executor로는 비동기 전용 executor
3. 알림 전송 과정에서 별도의 트랜잭션이 필요한 점 + 외부 서비스가 커밋된 이후 이벤트가 발행되어야 한다는 점으로 `@TransactionalEventListener` 적용
4. 파이어베이스 알림 전송 - 파이어베이스의 응답을 비동기적으로 받기 위해 `ApiFuture<String> process`에게 `addListener` 메서드 실행하도록 함 (n번째 파이어베이스 스레드가 n번째 분리된 스레드에게 응답을 담도록 하지 않도록)

알림 재전송 실패 시 지금은 단순히 로그를 남기는 방식으로 작성했는데, 이 부분이 괜찮을지는 논의해봐야 할 것 같습니다.

sosow0212 reviewed (changes requested):
수고하셨습니다!
다만 몇가지 질문할 것이 있어서 리뷰 남겼습니다!
해당 부분들 확인해주시고 코멘트 주시면, 확인하고 어프루브 하겠습니다!

eom-tae-in commented on src/main/java/com/atwoz/alert/domain/vo/AlertGroup.java:
```diff
@@ -0,0 +1,32 @@
+    INTERVIEW("인터뷰"),
+    ALERT("알림");
+
+    private String name;
```
final 키워드 붙여주세요~

eom-tae-in commented on src/main/java/com/atwoz/alert/domain/vo/AlertMessage.java:
```diff
@@ -0,0 +1,24 @@
+    private String title;
+    private String body;
+
+    private AlertMessage(final String title, final String body) {
+        this.title = title;
+        this.body = body;
+    }
```
이 생성자도 @AllArgsConstructor(access = AccessLevel.PRIVATE)로 하는게 어떨까요?

devholic22 replied:
동의합니다! AllArgsConstructor를 충분히 사용할 수 있었네요. 적용 완료했습니다

eom-tae-in commented on src/main/java/com/atwoz/alert/domain/Alert.java:
```diff
@@ -0,0 +1,77 @@
+    @Column(nullable = false)
+    private Long receiverId;
+
+    private Alert(final AlertGroup alertGroup, final AlertMessage alertMessage, final Long receiverId) {
+        this.alertGroup = alertGroup;
+        this.alertMessage = alertMessage;
+        this.receiverId = receiverId;
+        this.isRead = false;
+    }
+
+    public static Alert createWith(final AlertGroup group, final String title, final String body, final Long receiverId) {
+        AlertMessage message = AlertMessage.createWith(title, body);
+        return new Alert(group, message, receiverId);
+    }
```
이 로직을 @Builder를 추가해서 구현하는 건 어떨까요?

devholic22 replied:
빌더를 이미 사용할 수 있었네요! 생성자 코드도 제거될 수 있는 만큼 수정 완료했습니다

eom-tae-in commented on src/main/java/com/atwoz/alert/domain/AlertManager.java:
```diff
@@ -0,0 +1,6 @@
+package com.atwoz.alert.domain;
+
+public interface AlertManager {
+
+    void send(Alert alert, String sender, String token);
+}
```
AlertManager가 푸시 알람을 보내는 역할을 수행하는거 같은데 이게 도메인에 상태를 바꾸거나 연관이 있나요? 없다면 응용 계층에 두는게 맞을 것 같습니다!

devholic22 replied:
말씀하신대로 도메인과 관계없는 것 같아 응용 계층으로 수정 완료하였습니다!

eom-tae-in commented on src/main/java/com/atwoz/alert/domain/Alert.java:
```diff
@@ -0,0 +1,77 @@
+    @Column(nullable = false)
+    private Boolean isRead;
+
+    @Column(nullable = false)
+    @Enumerated
+    private AlertGroup alertGroup;
+
+    @Embedded
+    @Column(nullable = false)
+    private AlertMessage alertMessage;
```
어노테이션 순서 일치시켜주세요~ :)

devholic22 replied:
수정하였습니다!

eom-tae-in commented on src/main/java/com/atwoz/alert/exception/AlertExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class AlertExceptionHandler {
+
+    @ExceptionHandler(AlertSendException.class)
+    public ResponseEntity<String> handleAlertSendException(final AlertSendException e) {
+        return getExceptionWithStatus(e, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    @ExceptionHandler(ReceiverTokenNotFoundException.class)
+    public ResponseEntity<String> handleReceiverTokenNotFoundException(final ReceiverTokenNotFoundException e) {
+        return getExceptionWithStatus(e, HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(AlertNotFoundException.class)
+    public ResponseEntity<String> handleAlertNotFoundException(final AlertNotFoundException e) {
+        return getExceptionWithStatus(e, HttpStatus.NOT_FOUND);
+    }
+
+    private ResponseEntity<String> getExceptionWithStatus(final Exception exception, final HttpStatus status) {
+        return ResponseEntity.status(status)
+                .body(exception.getMessage());
+    }
+}
```
예외 상태 및 메세지를 담당하는 기능 재사용하신거 너무 좋습니다!

eom-tae-in reviewed (changes requested):
고생하셨습니다! 몇가지만 수정 부탁드릴게요~

devholic22 committed:
refactor: final 추가

- AlertGroup final 추가

devholic22 committed:
refactor: AllArgsConstructor 추가

- AlertMessage AllArgsConstructor 추가

devholic22 committed:
refactor: 어노테이션 순서 통일

- 어노테이션 순서 통일되도록 수정

devholic22 committed:
refactor: AlertManager 패키지 위치 수정

- AlertManager 인터페이스의 위치를 도메인에서 애플리케이션으로 변경

devholic22 committed:
refactor: Alert 생성 시 빌더 패턴 적용

- Alert 생성 시 빌더 패턴 적용되도록 수정

devholic22 committed:
refactor: AlertService orElseThrow 축약

- AlertService orElseThrow 축약 적용

devholic22 committed:
chore: Firebase 라이브러리 버전 업그레이드

- 파이어베이스 이후 버전의 특징 (sendEach, 트래픽 처리 개선 등)을 적용하기 위해 버전 업그레이드

devholic22 committed:
refactor: 알림의 특성을 고려하여 트랜잭션 분리 적용

- 외부 트랜잭션이 커밋된 이후에만 이벤트를 발송할 수 있도록 & 추가적인 데이터베이스 작업이 필요하므로 TransactionalEventListener로 변경
- 커밋 후 트랜잭션 작업을 하기 위해 REQUIRES_NEW 적용

devholic22 committed:
fix: AlertMessage Equals&HashCode 추가

- 테스트 정상화를 위해 AlertMessage Equals&HashCode 추가

devholic22 committed:
fix: AuditingHandler 추가 방식 이용

- AlertQueryRepositoryTest에 AuditingHandler 추가

devholic22 committed:
fix: LocalDateTime 비교 제외 진행

- 값 비교 시 LocalDateTime 제외되도록 수정

devholic22 committed:
refactor: AuditingHandler 제거

- AuditingHandler 제거

devholic22 committed:
refactor: AlertGroup Enumerated 수정

- AlertGroup Enumerated value 수정

devholic22 committed:
Merge branch 'develop' into feat/34

devholic22 committed:
Merge branch 'develop' into feat/34

devholic22 committed:
fix: import 교체

- 바뀐 경로 import 교체
- assertSoftly 문제 수정

devholic22 committed:
feat: 알림 삭제 스케줄러 분산 락 적용

- 레디스 분산 락을 적용하기 위해 redisson 등록
- AlertScheduler 인터페이스 분리
- 기존 AlertService 스케줄러 메서드 삭제 및 이동
- RedissonAlertSchedulerTest 테스트 작성

devholic22 committed:
Merge branch 'feat/34' of https://github.com/sosow0212/atwoz into feat/34

devholic22 committed:
refactor: Alert 인덱싱 적용

devholic22 committed:
refactor: 비동기 적용 방식 수정

- Async 어노테이션으로 알림 전송 스레드 분리
- Async 전용 설정 파일 생성
- Firebase 전용 스레드 매니저 생성
- 트랜잭션 전파 수준 수정

sosow0212 commented on src/main/java/com/atwoz/alert/config/FirebaseThreadManager.java:
```diff
@@ -0,0 +1,30 @@
+@Component
+public class FirebaseThreadManager extends ThreadManager {
+
+    private static final int FIREBASE_THREADS_SIZE = 40;
```
쓰레드 40개로 선정하신 이유가 궁금합니다!

devholic22 replied:
사실 이 부분은 스레드를 얼마까지 정해뒀을 때 수요를 버틸 수 있을지 정확히 예측하기 힘들다고 판단하여 다른 글들을 참고해서 보수적으로 적용했었습니다. 실제 배포 후 모니터링한 뒤 더 늘릴 수 있다면 이후에 늘려보고 싶습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/config/FirebaseThreadManager.java:
```diff
@@ -0,0 +1,30 @@
+
+    private static final int FIREBASE_THREADS_SIZE = 40;
+
+    @Override
+    protected ExecutorService getExecutor(final FirebaseApp firebaseApp) {
+        return Executors.newFixedThreadPool(FIREBASE_THREADS_SIZE);
+    }
```
newFixedThreadPool로 쓰레드를 설정해주셨는데요.
만약에 40개의 코어 쓰레드가 모두 사용중이라면 그 이후에 오는 요청은 어떻게 되고 최대 몇 개까지 기다릴 수 있을까요?
그리고 만약 초과가 됐다면 이 경우 대기중인 요청은 어떻게 되나요?

devholic22 replied:
* 우선 알림을 보낼 때 영향이 끼쳐지는 스레드는 FirebaseThreadManager에서 관리하는 스레드가 아닌, AsyncConfig에서 관리하는 스레드입니다. (AsyncConfig의 getAsyncExecutor 메서드, 스레드 이름을 로그로 남겨보면서 확인하였습니다.)
  
![스크린샷 2024-07-28 오후 5 17 01](https://github.com/user-attachments/assets/4345a63b-00a6-4a35-b6d0-3c3c8b850fe4)

* 지정한 `MAX_THREAD_SIZE + QUEUE_SIZE`개 만큼의 수요 (지금은 MAX_THREAD_SIZE 40, QUEUE_SIZE 100이므로 140개 가능)를 감당한 뒤 나머지 처리에 대해서는 ThreadPoolExecutor에서 전략을 설정할 수 있는데, 우선은 이벤트를 호출한 스레드 (즉, 스프링 메인 스레드)에서 나머지를 보내도록 설정하였습니다. (ThreadPoolExecutor.CallerRunsPolicy 전략)
  * `AbortPolicy`: TaskRejectedException이 발생되고 요청이 무시됩니다.
  * `CallerRunsPolicy`: 처리되지 못한 요청을 ThreadPool을 호출한 Thread에서 처리합니다.
  * `DiscardPolicy`: 처리되지 못한 요청을 무시합니다.
  * `DiscardOldestPolicy`: 가장 오래된 요청을 삭제하고 다시 execute를 실행합니다.
  * 이에 따라, 알림을 보내야 하는 상황에서 무시하거나 예외를 발생시키기보다는 메인 스레드를 이용해서라도 진행시켜야 하는 게 맞겠다는 생각이 들어 CallerRunsPolicy를 선택하였습니다.
* 그럼에도 FirebaseThreadManager에서 스레드를 따로 지정한 이유는 기본적으로 파이어베이스가 요청이 n개 들어온다면 n개만큼의 스레드를 만들기 때문에, 극단적인 예로 10,000개의 알림을 보내야 한다면 10,000개의 스레드를 만들게 되고 이로 인해 OutOfMemory 현상이 발생할 수 있음을 확인하였어서 지정했었습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/application/AlertQueryService.java:
```diff
@@ -0,0 +1,34 @@
+        return AlertPagingResponse.of(response, nextPage);
+    }
+
+    private int getNextPage(final int pageNumber, final Page<AlertSearchResponse> alerts) {
+        if (alerts.hasNext()) {
+            return pageNumber + NEXT_PAGE_INDEX;
+        }
+        return NO_MORE_PAGE;
+    }
+}
```
이 부분은 페이징을 사용할 때 중복 코드로 발생할 것 같은데
하나의 템플릿을 제네릭으로 만들고 페이징할 때 그걸 extends해서 사용하면 어떨까요?

devholic22 replied:
좋은 생각입니다! `BaseQueryService` 템플릿을 만들어두었습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/FirebaseAlertManager.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public void send(final Alert alert, final String sender, final String token) {
+        Message firebaseMessage = createAlertMessage(alert, sender, token);
+        FirebaseMessaging firebase = FirebaseMessaging.getInstance();
+        ApiFuture<String> process = firebase.sendAsync(firebaseMessage);
+
+        Runnable task = () -> logAlertResult(process, firebaseMessage);
+        process.addListener(task, executor);
+    }
```
AlertManager.send()를 사용하는 부분에서 firebase는 외부 라이브러리일텐데 해당 메서드에서 firebase를 통해 알람을 전송하면 firebase 내부적으로 비동기로 작동이 되나요? (즉, send 메서드에서 firebase에서 알림을 전송할 때 반환을 받든 안 받든 상관없이 기다리지 않냐는 질문입니다!)

devholic22 replied:
* firebase의 sendAsync 메서드를 이용하면 파이어베이스 내부적으로도 비동기적으로 작동되고, ApiFuture (Future의 자식 객체)를 반환합니다.
* 이때 ApiFuture.get 메서드를 이용하면 응답 값을 받을 수 있으나, get 메서드를 이용하는 순간 동기적으로 작동하기 때문에 get을 통한 방식보다는 addListener 방식을 이용하여 콜백 메서드로 실행되게끔 하였습니다. (비동기-논블록킹)

![스크린샷 2024-07-28 오후 7 46 41](https://github.com/user-attachments/assets/eacd3745-8f0e-4584-906e-39a83cb0e4b7)
예시로 첫 번째 경우는 `firebase.sendAsync(firebaseMessage).get()` 방식 (동기)으로 바로 이끌어낸 경우이며, 이 때는 알림 작업을 진행한 스레드가 알림 작업의 결과까지 받아옵니다.

![스크린샷 2024-07-28 오후 7 51 36](https://github.com/user-attachments/assets/ed0cc336-ad73-47fc-a592-be77f06d506d)
그러나 두 번째 경우 (현재 코드, Runnable 및 addListener 방식)는 알림 작업을 진행한 스레드가 결과를 받는 게 아니라, 스레드 풀에 남아있는 다른 스레드가 알림 작업의 결과를 받아오게 됩니다.

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/FirebaseAlertManager.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private void retryAlert(final Message message, final ExecutionException exception) {
+        if (exception.getCause() instanceof FirebaseMessagingException e) {
```
instanceof는 지양해주세요! [참고](https://tecoble.techcourse.co.kr/post/2021-04-26-instanceof/)

devholic22 replied:
instanceof 대신 ClassCastException을 try-catch 하는 식으로 리팩터링 하였습니다.

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/FirebaseAlertManager.java:
```diff
@@ -0,0 +1,132 @@
+        return errorCode.equals(MessagingErrorCode.INTERNAL) || errorCode.equals(MessagingErrorCode.UNAVAILABLE);
+    }
+
+    private void retryInThreeTimes(final Message message) {
```
많은 양의 알람을 재전송하는데 firebase측의 문제로 쓰레드가 오랫동안 작업을 수행한다면 문제가 생길 수도 있을 것 같은데, redis에 적재해놓고 한 번에 배치처리 하는 것은 어떻다고 생각하시나요?

devholic22 replied:
파이어베이스 알림 전송 과정에서 재전송을 해야 하는 경우가 빈번히 발생하지는 않을 것 같아 지금은 배치 처리를 하지 않았는데, 지금 단계에서 배치 작업을 진행해야 할 지 고민이 듭니다. 모니터링 과정을 거친 뒤에 실제 알림이 빈번히 재전송을 해야 할 필요성이 느껴질 때 반영해도 괜찮을까요?

sosow0212 commented on src/main/java/com/atwoz/alert/infrastructure/RedissonAlertScheduler.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Scheduled(cron = MIDNIGHT)
+    @Override
+    public void deleteExpiredAlerts() {
```
네임드락 공통 코드 부분도 AOP로 만들어주시면 재사용하기에 도움될 것 같습니다!

devholic22 replied:
`@RedissonDistributedLock`로 AOP 등록 완료하였습니다!

sosow0212 reviewed (changes requested):
몇 부분에 대해 질문이 있어서 남겨봤습니다 확인 부탁드립니다!

devholic22 committed:
refactor: QueryService abstract class 생성

- 페이지 조회 공통 메서드를 축약하기 위해 BaseQueryService 생성

devholic22 committed:
refactor: RedissonLock AOP 생성

- Redisson Lock 재사용을 위해 AOP 생성 및 적용

devholic22 committed:
refactor: 비동기 예외 핸들러 등록

- AsyncUncaughtExceptionHandler 등록

devholic22 committed:
refactor: Alert 생성 과정 수정

- DB에 저장된 createdAt 속성 들어가도록 수정
- 메시지 바디 null 문제로 인해 data에 들어가도록 수정

devholic22 committed:
refactor: instanceof 표현 삭제

- instanceof 대신 ClassCastException 탐지되도록 수정
